src/fmtspec.jl

```diff
@@ -164,25 +164,42 @@ _srepr(x) = repr(x)
 _srepr(x::AbstractString) = x
 _srepr(x::AbstractChar) = string(x)
 _srepr(x::Enum) = string(x)
+@static if VERSION < v"1.2.0-DEV"
+    _srepr(x::Irrational{sym}) where {sym} = string(sym)
+end
 
 function printfmt(io::IO, fs::FormatSpec, x)
     cls = fs.cls
     ty = fs.typ
     if cls == 'i'
-        ix = Integer(x)
-        ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, ix, _Dec()) :
-        ty == 'x' ? _pfmt_i(io, fs, ix, _Hex()) :
-        ty == 'X' ? _pfmt_i(io, fs, ix, _HEX()) :
-        ty == 'o' ? _pfmt_i(io, fs, ix, _Oct()) :
-        _pfmt_i(io, fs, ix, _Bin())
+        try
+            ix = Integer(x)
+            ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, ix, _Dec()) :
+            ty == 'x' ? _pfmt_i(io, fs, ix, _Hex()) :
+            ty == 'X' ? _pfmt_i(io, fs, ix, _HEX()) :
+            ty == 'o' ? _pfmt_i(io, fs, ix, _Oct()) :
+            _pfmt_i(io, fs, ix, _Bin())
+        catch
+            ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, x, _Dec()) :
+            ty == 'x' ? _pfmt_i(io, fs, x, _Hex()) :
+            ty == 'X' ? _pfmt_i(io, fs, x, _HEX()) :
+            ty == 'o' ? _pfmt_i(io, fs, x, _Oct()) :
+            _pfmt_i(io, fs, x, _Bin())
+        end
     elseif cls == 'f'
-        fx = float(x)
-        if isfinite(fx)
-            ty == 'f' || ty == 'F' ? _pfmt_f(io, fs, fx) :
-            ty == 'e' || ty == 'E' ? _pfmt_e(io, fs, fx) :
+        try
```

Comment:
Same as above - if it can't get converted to something that is `AbstractFloat`, then it should output as string, or give an error.